library/std/src/sync/reentrant_lock.rs

```diff
@@ -72,14 +73,95 @@ use crate::sys::sync as sys;
 // since we're not dealing with multiple threads. If it's not equal,
 // synchronization is left to the mutex, making relaxed memory ordering for
 // the `owner` field fine in all cases.
+//
+// On systems without 64 bit atomics we use a simple seqlock to emulate a 64 bit Tid using
+// 32 bit atomics (which should be supported on all platforms with `std`). This works
+// because only one thread at a time (the one holding the mutex) writes to it.
 #[unstable(feature = "reentrant_lock", issue = "121440")]
 pub struct ReentrantLock<T: ?Sized> {
     mutex: sys::Mutex,
-    owner: AtomicUsize,
+    owner: Tid,
     lock_count: UnsafeCell<u32>,
     data: T,
 }
 
+cfg_if!(
+    if #[cfg(target_has_atomic = "64")] {
+        use crate::sync::atomic::{AtomicU64, Ordering::Relaxed};
+
+        struct Tid(AtomicU64);
+
+        impl Tid {
+            const fn new(tid: u64) -> Self {
+                Self(AtomicU64::new(tid))
+            }
+
+            #[inline]
+            fn get(&self) -> u64 {
+                self.0.load(Relaxed)
+            }
+
+            #[inline]
+            fn set(&self, tid: u64) {
+                self.0.store(tid, Relaxed)
+            }
+        }
+    } else if #[cfg(target_has_atomic = "32")] {
+        use crate::sync::atomic::{AtomicU32, Ordering::{Acquire, Relaxed, Release}};
+
+        struct Tid {
+            seq: AtomicU32,
+            low: AtomicU32,
+            high: AtomicU32,
+        }
+
+        impl Tid {
+            const fn new(tid: u64) -> Self {
```

Comment:
This is subjective, but for me this code would be more readable if the `ThreadId` abstraction was only taken away inside here. Upon first glance, this PR seemed unsound to me because I didn't remember that `ThreadId`s are guaranteed to be non-zero, using `u64` instead of `Option<ThreadId>` in the comparison makes the logic clearer. May I suggest doing
```suggestion
            const fn new(tid: Option<ThreadId>) -> Self {
```
and so on?